src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1207,6 +1217,59 @@ private Task WriteStringAsync(string s, bool async)
             return WriteStringAsyncSlow(s, async);
         }
 
+        private Task WriteStringAsync(string s, Encoding? encoding, bool async)
+        {
+            if (encoding is null)
+            {
+                return WriteStringAsync(s, async);
+            }
+
+            // If there's enough space in the buffer to just copy all of the string's bytes, do so.
+            int offset = _writeOffset;
+            int available = _writeBuffer.Length - offset;
+
+            if (encoding.IsSingleByte ? available >= s.Length : (ReferenceEquals(encoding, Encoding.UTF8) && available >= s.Length * 3))
+            {
+                int written = encoding.GetBytes(s, _writeBuffer.AsSpan(offset));
+                Debug.Assert(written == s.Length || (ReferenceEquals(encoding, Encoding.UTF8) && written <= s.Length * 3));
+                _writeOffset = offset + written;
+                return Task.CompletedTask;
+            }
+
+            // Otherwise, fall back to doing a normal slow string write
+            return WriteStringWithEncodingAsyncSlow(s, encoding, async);
+        }
+
+        private async Task WriteStringWithEncodingAsyncSlow(string s, Encoding encoding, bool async)
+        {
+            int length;
+
+            if (encoding.IsSingleByte)
+            {
+                length = s.Length;
+            }
+            else if (ReferenceEquals(encoding, Encoding.UTF8) && s.Length <= 1024)
+            {
+                // Avoid calculating the length if the rented array would be small anyway
+                length = s.Length * 3;
```

Comment:
It's a lot clearer this way, though slightly more overhead. This is a `Slow` method, so I'd give preference to clarity.
```suggestion
                length = Encoding.UTF8.GetMaxByteCount(s.Length);
```

Reply:
Simplified the slow path to
```c#
int length = s.Length <= 512
    ? encoding.GetMaxByteCount(s.Length)
    : encoding.GetByteCount(s);
```

I also changed the fast path to
```c#
const int MaxUtf8BytesPerChar = 3;
bool sufficientBufferSpace = ReferenceEquals(encoding, Encoding.Latin1) ? available >= s.Length :
    ReferenceEquals(encoding, Encoding.UTF8) ? available >= s.Length * MaxUtf8BytesPerChar :
    available >= encoding.GetMaxByteCount(s.Length);
```

Meaning we don't have to always go through the ArrayPool for custom encodings